skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,433 @@
+            centering (bool): if True then calculate the mean of the functional
+                data object and center the data first. Defaults to True
+        """
+        self.n_components = n_components
+        self.centering = centering
+
+    @abstractmethod
+    def fit(self, X, y=None):
+        """Computes the n_components first principal components and saves them
+        inside the FPCA object.
+
+        Args:
+            X (FDataGrid or FDataBasis):
+                the functional data object to be analysed
+            y (None, not used):
+                only present for convention of a fit function
+
+        Returns:
+            self (object)
+        """
+        pass
+
+    @abstractmethod
+    def transform(self, X, y=None):
+        """Computes the n_components first principal components score and
+        returns them.
+
+        Args:
+            X (FDataGrid or FDataBasis):
+                the functional data object to be analysed
+            y (None, not used):
+                only present because of fit function convention
+
+        Returns:
+            (array_like): the scores of the data with reference to the
+            principal components
+        """
+        pass
 
-    It uses SVD numpy implementation to compute PCA.
+    def fit_transform(self, X, y=None, **fit_params):
+        """Computes the n_components first principal components and their scores
+        and returns them.
+        Args:
+            X (FDataGrid or FDataBasis):
+                the functional data object to be analysed
+            y (None, not used):
+                only present for convention of a fit function
 
-    Args:
-        fdatagrid (FDataGrid): functional data object.
-        n (int, optional): Number of principal components. Defaults to 2.
+        Returns:
+            (array_like): the scores of the data with reference to the
+            principal components
+        """
+        self.fit(X, y)
+        return self.transform(X, y)
 
-    Returns:
-        tuple: (scores, principal directions, eigenvalues)
+
+class FPCABasis(FPCA):
+    """Funcional principal component analysis for functional data represented
+    in basis form.
+
+    Attributes:
+        n_components (int): number of principal components to obtain from
+            functional principal component analysis. Defaults to 3.
+        centering (bool): if True then calculate the mean of the functional data
+            object and center the data first. Defaults to True. If True the
+            passed FDataBasis object is modified.
+        components_ (FDataBasis): this contains the principal components either
+            in a basis form.
+        components_basis (Basis): the basis in which we want the principal
+            components. We can use a different basis than the basis contained in
+            the passed FDataBasis object.
+        component_values_ (array_like): this contains the values (eigenvalues)
```

Comment:
Why not `explained_variance_`, as in scikit-learn?

Reply:
Extra computation is needed to obtain the principal components. Maybe we can expose pca_ so the user can choose what he wants to obtain from this module? As pca from scikit learn provides more attributes that the user may be interested in.

Reply:
I think that explained_variance_ratio is better, as it gives the percentage which is more interesting

Reply:
Why not both?

Reply:
Yeah it costs nothing and both may be interesting

Reply:
This still has a different name

Reply:
What is the difference between `component_values_ ` and `explained_variance_ `?

Reply:
The difference is that explained_variance_ = (singular_values ** 2) / (n_samples - 1)

where singular values**2 = component_values

I have removed component_values as it is redundant in this case.